devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy

```diff
@@ -82,113 +83,159 @@ private List<Map> collectAttributes(Map descMap) {
-    if (lastLevel == "undefined" || currentLevel == lastLevel) { //Ignore invalid reports
-        log.debug "Ignore invalid reports"
-    } else {
-        sendEvent(name: "level", value: currentLevel)
-        if (currentLevel == 0 || currentLevel == 100) {
-            sendEvent(name: "windowShade", value: currentLevel == 0 ? "closed" : "open")
-        } else {
-            if (lastLevel < currentLevel) {
-                sendEvent([name:"windowShade", value: "opening"])
-            } else if (lastLevel > currentLevel) {
-                sendEvent([name:"windowShade", value: "closing"])
-            }
-            runIn(1, "updateFinalState", [overwrite:true])
-        }
-    }
+	def lastLevel = device.currentValue("level")
+	log.debug "levelEventHandle - currentLevel: ${currentLevel} lastLevel: ${lastLevel}"
+	if (lastLevel == "undefined" || currentLevel == lastLevel) { //Ignore invalid reports
+		log.debug "Ignore invalid reports"
+	} else {
+		sendEvent(name: "level", value: currentLevel)
+		if (currentLevel == 0 || currentLevel == 100) {
+			sendEvent(name: "windowShade", value: currentLevel == 0 ? "closed" : "open")
+		} else {
+			if (lastLevel < currentLevel) {
+				sendEvent([name:"windowShade", value: "opening"])
+			} else if (lastLevel > currentLevel) {
+				sendEvent([name:"windowShade", value: "closing"])
+			}
+			runIn(1, "updateFinalState", [overwrite:true])
+		}
+	}
 }
 
 def updateFinalState() {
-    def level = device.currentValue("level")
-    log.debug "updateFinalState: ${level}"
-    if (level > 0 && level < 100) {
-        sendEvent(name: "windowShade", value: "partially open")
-    }
+	def level = device.currentValue("level")
+	log.debug "updateFinalState: ${level}"
+	if (level > 0 && level < 100) {
+		sendEvent(name: "windowShade", value: "partially open")
+	}
 }
 
 def supportsLiftPercentage() {
-    device.getDataValue("manufacturer") != "Feibit Co.Ltd"
+	device.getDataValue("manufacturer") != "Feibit Co.Ltd"
 }
 
 def close() {
-    log.info "close()"
-    zigbee.command(CLUSTER_WINDOW_COVERING, COMMAND_CLOSE)
+	log.info "close()"
+	zigbee.command(CLUSTER_WINDOW_COVERING, COMMAND_CLOSE)
 }
 
 def open() {
-    log.info "open()"
-    zigbee.command(CLUSTER_WINDOW_COVERING, COMMAND_OPEN)
+	log.info "open()"
+	zigbee.command(CLUSTER_WINDOW_COVERING, COMMAND_OPEN)
 }
 
 def setLevel(data, rate = null) {
-    log.info "setLevel()"
-    def cmd
-    if (supportsLiftPercentage()) {
-        cmd = zigbee.command(CLUSTER_WINDOW_COVERING, COMMAND_GOTO_LIFT_PERCENTAGE, zigbee.convertToHexString(data, 2))
-    } else {
-        cmd = zigbee.command(zigbee.LEVEL_CONTROL_CLUSTER, COMMAND_MOVE_LEVEL_ONOFF, zigbee.convertToHexString(Math.round(data * 255 / 100), 2))
-    }
+	log.info "setLevel()"
+	def cmd
+	if (supportsLiftPercentage()) {
+		if(isIkeaKadrilj()){
+			// Ikea Kadrilj keeps % level of being closed (instead of % level of being opened)
+			// inverting that logic is needed here
+			def newData = 100 - data
```

Comment:
just overwrite `data` and have one `command` call

Reply:
ok, done